src/node_file.cc

```diff
@@ -2727,6 +2730,222 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
+  std::string_view hostname = file_url->get_hostname();
+
+  if (hostname.size() > 0) {
+    std::string error_message =
+        std::string("File URL host must be \"localhost\" or empty on ") +
+        std::string(per_process::metadata.platform);
+    env->isolate()->ThrowException(
+        ERR_INVALID_FILE_URL_HOST(env->isolate(), error_message.c_str()));
+
+    return false;
+  }
+
+  int first_percent = 0;
+
+  for (std::string::size_type i = 0; (i + 2) < pathname.size(); ++i) {
+    if (pathname[i] != '%') continue;
+
+    if (first_percent == 0) {
+      first_percent = i;
+    }
+
+    if (pathname[i + 1] != '2' || (pathname[i + 2] | 0x20) != 102) continue;
+
+    env->isolate()->ThrowException(ERR_INVALID_FILE_URL_PATH(
+        env->isolate(), "File URL path must not include encoded / characters"));
+
+    return false;
+  }
+
+  *result_file_path = ada::unicode::percent_decode(pathname, pathname.size());
+
+  return true;
+#endif
+}
+
+FilePathIsFileReturnType FilePathIsFile(Environment* env, const std::string& file_path) {
+  THROW_IF_INSUFFICIENT_PERMISSIONS(
+      env, permission::PermissionScope::kFileSystemRead, file_path, FilePathIsFileReturnType::kThrowInsufficientPermissions);
+
+  uv_fs_t req;
+
+  int rc = uv_fs_stat(env->event_loop(), &req, file_path.c_str(), nullptr);
+
+  if (rc == 0) {
+    const uv_stat_t* const s = static_cast<const uv_stat_t*>(req.ptr);
+    rc = !!(s->st_mode & S_IFDIR);
+  }
+
+  uv_fs_req_cleanup(&req);
+
+  // rc is 0 if the path refers to a file
+  if (rc == 0)
+    return FilePathIsFileReturnType::kIsFile;
+
+  return FilePathIsFileReturnType::kIsNotFile;
+}
+
+static void LegacyMainResolve(const FunctionCallbackInfo<Value>& args) {
+  CHECK_GE(args.Length(), 1);
+  // packageJSONUrl
+  // /home/user/project/package.json
+  CHECK(args[0]->IsString());
+
+  Environment* env = Environment::GetCurrent(args);
+
+  std::string_view base_path =
+      Utf8Value(env->isolate(), args[0].As<String>()).ToStringView();
+  auto base = ada::parse<ada::url_aggregator>(base_path);
+
+  if (!base) {
+    return args.GetReturnValue().Set(LegacyMainResolveReturnType::kInvalidUrl);
+  }
+
+  ada::url_aggregator* base_pointer = &base.value();
+  ada::url_aggregator* file_path_agg;
+  ada::result<ada::url_aggregator> file_path_url;
+  std::unique_ptr<std::string> initial_file_path;
+  std::string file_path;
+
+  // packageConfig.main
+  // ./index.js
+  // ./lib/index.js
+  if (args[1]->IsString()) {
+    std::string package_config_main =
+        Utf8Value(env->isolate(), args[1].As<String>()).ToString();
+
+    file_path_url = ada::parse<ada::url_aggregator>(
+        std::string("./") + package_config_main, base_pointer);
+
+    if (!file_path_url) {
+      return args.GetReturnValue().Set(LegacyMainResolveReturnType::kInvalidUrl);
+    }
+
+    file_path_agg = &file_path_url.value();
+
+    if (!FileURLToPathImpl(env, file_path_agg, initial_file_path)) return;
+
+    FromNamespacedPath(&(*initial_file_path));
```

Comment:
I recommend dropping pointer and just pass the reference:

```suggestion
    FromNamespacedPath(initial_file_path);
```

Reply:
Since FromNamespacedPath() mutates the string, it can't be a const ref, so I think we should still pass a pointer (it's mentioned in the c++ style guide). If this is for readability, what about using get():
```suggestion
    FromNamespacedPath(initial_file_path.get());
```

Reply:
Instead of using `unique_ptr`, I change to `std::string&` on the method and this line became `FromNamespacedPath(&initial_file_path);`, much simpler.